rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,856 @@
+
+## 3. Standardize Components
+[guide-standardization]: #3-standardize-components
+
+_Don't repeat yourself_
+
+CTD's applications largely share a common architecture, which in turn means that most applications use the same types of AWS resources. Specifically: Most of our applications run as ECS services, many include a load balancer, and some also store data in an RDS database. These components largely have common configuration, save for a few attributes that need to be customized for each application.
+
+Recently, effort has been put toward defining Terraform modules that encapsulate as much of the configuration as possible for these infrastructure components. Those modules are:
+
+- `ctd-ecs-app` for ECS service and load balancer configuration
+- `ctd-rds-db` for RDS database configuration
+
+These _infrastructure component modules_ have seen some adoption, but are not yet widely used. To maximize their effectiveness, we will:
+
+- Set up a testing workflow for each module using the process described in the [Testing Workflow][guide-module-testing] section
+- Move them to separate repos and version them, using the process described in the [Version and Publish Modules section][guide-module-versioning]
+
+We will also push to migrate all applications to use these modules to minimize redundant and nonstandard configuration.
+
+Relatedly, the `ctd-ecs-app` module introduced the concept of a `context` var that bundles all baseline configuration need by a module, including VPC and subnet information, and shared resources such as TLS certificates and Cognito domains. This convention has the potential to significantly reduce redundant configuration in app modules, and should also be adopted widely.
+
+See the corresponding [reference-level explanation][reference-standardization] for related recommendations.
+
+## 4. Version and Publish Modules
+[guide-module-versioning]: #4-version-and-publish-modules
+
+_Modules as immutable artifacts_
+
+Nearly all of CTD's modules are currently stored in the devops repo and called using a relative path to the module's directory. Terraform supports a number of alternate methods for sourcing modules, including:
+
+- Git repositories (including specific refs)
+- Public module registries
+- Private module registries
+
+The main advantage of using a repo or module registry instead of a path is that the module can then be treated as a versioned artifact. If we are able to pin module calls to a specific version, we can make changes to the module without immediately impacting all the contexts in which it's called. This is useful for both infrastructure component modules, which are used across many different applications, and application modules, which are used to build multiple distinct environments. It also reduces the need to rely on feature gating mechanisms such as module variables, which is the typical approach to module changes currently.
```

Comment:
I agree that this is a potential advantage, but it's also a tradeoff in that it sets up another queue of version upgrades that need to be handled periodically so that configuration doesn't end up relying on a stale version when an urgent fix is made. I think it makes more sense to version the widely-shared standard modules than application modules, but even then it still raises this issue.